Civi/Civioffice/DocumentEditorType/CollaboraOnlineEditorType.php

```diff
@@ -0,0 +1,216 @@
+      && filter_var($wopiSrcHostname, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME) === FALSE
+    ) {
+      $form->setElementError('wopi_src_hostname', E::ts('Invalid value'));
+    }
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function postProcessSettingsForm(\CRM_Civioffice_Form_DocumentEditorSettings $form): array {
+    $values = $form->exportValues();
+
+    return [
+      'cool_url' => $values['cool_url'],
+      'wopi_src_hostname' => $values['wopi_src_hostname'] ?? NULL,
+    ];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getDefaultConfiguration(): array {
+    return [];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getWopiDiscoveryUrl(array $configuration): string {
+    assert(is_string($configuration['cool_url']));
+
+    return rtrim($configuration['cool_url'], '/') . '/hosting/discovery';
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function isFileSupported(array $configuration, array $file, int $editorId): bool {
+    return NULL !== $this->getWopiAppUrl($file, $editorId);
+  }
+
+  /**
+   * @inheritDoc
+   *
+   * @throws \Psr\Http\Client\ClientExceptionInterface
+   */
+  public function handleFile(array $configuration, array $file, int $editorId): Response {
+    $wopiAppUrl = $this->getWopiAppUrl($file, $editorId);
+    assert(NULL !== $wopiAppUrl);
+
+    $fileBaseName = basename($file['uri']);
+    $wopiSrc = $this->civiUrlGenerator->generateAbsoluteUrl('civicrm/civioffice/collabora/wopi/files/' . $file['id']);
+    if ('' !== ($configuration['wopi_src_hostname'] ?? '')) {
+      assert(is_string($configuration['wopi_src_hostname']));
+      $wopiSrcHostname = parse_url($wopiSrc, PHP_URL_HOST);
+      assert(is_string($wopiSrcHostname));
+      $wopiSrc = str_replace($wopiSrcHostname, $configuration['wopi_src_hostname'], $wopiSrc);
+    }
+
+    $wopiUrl = rtrim($wopiAppUrl, '?') . '?WOPISrc=' . urlencode($wopiSrc);
+    $accessTokenTtl = \CRM_Utils_Time::time() + self::TOKEN_VALIDITY_TIME;
+    $accessToken = $this->accessTokenService->generateToken(
+      $file['id'],
+      $editorId,
+      $accessTokenTtl
+    );
+    $accessTokenTtlMs = $accessTokenTtl * 1000;
+
+    $html = <<<HTML
+<html>
+<head>
+<title>$fileBaseName - CiviCRM</title>
+</head>
+<body style="margin: 0; overflow: hidden">
+
+<iframe id="coolframe" name="coolframe" title="Collabora Online"
+  allowfullscreen allow="clipboard-read *; clipboard-write *"
+  style="width:100%; height:100%; position:absolute; border: none;"></iframe>
+
+<div style="display: none">
+  <form id="coolform" name="coolform" target="coolframe" action="$wopiUrl" method="post">
+    <input name="access_token" value="$accessToken" type="hidden">
+    <input name="access_token_ttl" value="$accessTokenTtlMs" type="hidden">
+  </form>
+</div>
+
+<script>
+  const frame = document.getElementById("coolframe");
+  const form = document.getElementById("coolform");
+  frame.focus();
+  form.submit();
+</script>
+</body>
+</html>
+HTML;
```

Comment:
I understand running this through a template engine doesn't have any benefits, but seeing that much HTML, CSS and JS code in here kind of bothers me …
And what was the reason again this needs a form post (can't remember right now)?